File: src/main/java/com/baulsupp/oksocial/security/CertificateUtils.java
package com.baulsupp.oksocial.security;

import com.secdec.codedx.security.CompositeX509TrustManager;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import javax.net.ssl.TrustManagerFactory;
import javax.net.ssl.X509TrustManager;

public class CertificateUtils {
  public static X509TrustManager load(List<File> serverCerts)
      throws NoSuchAlgorithmException, KeyStoreException, IOException, CertificateException {
    return trustManagerForKeyStore(keyStoreForCerts(serverCerts));
  }

  public static X509TrustManager trustManagerForKeyStore(KeyStore ks)
      throws NoSuchAlgorithmException, KeyStoreException {
    TrustManagerFactory tmf =
        TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());

    tmf.init(ks);

    return (X509TrustManager) tmf.getTrustManagers()[0];
  }

  public static KeyStore keyStoreForCerts(List<File> serverCerts)
      throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException {
    CertificateFactory cf = CertificateFactory.getInstance("X.509");

    KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
    ks.load(null);

    for (int i = 0; i < serverCerts.size(); i++) {
      try (InputStream is = new FileInputStream(serverCerts.get(i))) {
        X509Certificate caCert = (X509Certificate) cf.generateCertificate(is);
        ks.setCertificateEntry("cacrt." + i, caCert);
      }
    }
    return ks;
  }

  public static X509TrustManager combineTrustManagers(List<X509TrustManager> trustManagers)
      throws NoSuchAlgorithmException, KeyStoreException, IOException, CertificateException {
    trustManagers.add(load(includedCertificates()));
    trustManagers.add(systemTrustManager());

    return new CompositeX509TrustManager(trustManagers);
  }

  public static X509TrustManager systemTrustManager()
      throws NoSuchAlgorithmException, KeyStoreException {
    return trustManagerForKeyStore(null);
  }

  public static List<File> includedCertificates() {
    String installDir = System.getenv("INSTALLDIR");
    if (installDir == null) {
      installDir = ".";
    }

    File[] files =
        new File(installDir, "certificates").listFiles(f -> f.getName().endsWith(".crt"));

    if (files != null) {
      return Arrays.asList(files);
    }

    return null;
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/security/CertificateUtils.kt
package com.baulsupp.oksocial.security

import java.io.File
import java.io.FileInputStream
import java.security.KeyStore
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager

object CertificateUtils {

  fun load(serverCerts: List<File>): X509TrustManager {
    return trustManagerForKeyStore(keyStoreForCerts(serverCerts))
  }

  fun trustManagerForKeyStore(ks: KeyStore?): X509TrustManager {
    val tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())

    tmf.init(ks)

    return tmf.trustManagers[0] as X509TrustManager
  }

  fun keyStoreForCerts(serverCerts: List<File>): KeyStore {
    val cf = CertificateFactory.getInstance("X.509")

    val ks = KeyStore.getInstance(KeyStore.getDefaultType())
    ks.load(null)

    for (i in serverCerts.indices) {
      FileInputStream(serverCerts[i]).use { `is` ->
        val caCert = cf.generateCertificate(`is`) as X509Certificate
        ks.setCertificateEntry("cacrt.$i", caCert)
      }
    }
    return ks
  }

  fun combineTrustManagers(trustManagers: MutableList<X509TrustManager>): X509TrustManager {
    val localCerts = includedCertificates()
    if (localCerts != null) {
      trustManagers.add(load(localCerts))
    }
    trustManagers.add(systemTrustManager())

    return MergedX509TrustManager(trustManagers)
  }

  fun systemTrustManager(): X509TrustManager {
    return trustManagerForKeyStore(null)
  }

  fun includedCertificates(): List<File>? {
    var installDir: String? = System.getenv("INSTALLDIR")
    if (installDir == null) {
      installDir = "."
    }

    return File(installDir, "certificates").listFiles { f -> f.name.endsWith(".crt") }?.toList()
  }
}
